File: src/stats/calculator.ts
/**
 * Stat calculator orchestrator
 * Calculates stat values and manages OBS text source updates
 */

import { DndBeyondCharacterResponse } from "../types.js";
import { CalculatedStat, ComputedStat, StatId, StatMapping } from "./types.js";
import { statDefinitions, getExpressionVariables } from "./definitions.js";
import { evaluateExpression } from "./expression.js";

/**
 * Calculator for D&D Beyond stats
 */
export class StatCalculator {
  /**
   * Calculate a single stat value
   * @throws Error if stat is unknown
   */
  calculateStat(statId: StatId, data: DndBeyondCharacterResponse): string | number {
    const definition = statDefinitions[statId];
    if (!definition) {
      throw new Error(`Unknown stat: ${statId}`);
    }

    try {
      return definition.calculate(data);
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      console.warn(`[STATS] Error calculating stat ${statId}: ${errorMessage}`);
      throw error;
    }
  }

  /**
   * Calculate all configured stat mappings
   * Returns array of { obsSourceName, value, changed } tuples
   */
  calculateMappings(
    mappings: StatMapping[],
    data: DndBeyondCharacterResponse,
    previousValues?: Map<string, string>
  ): CalculatedStat[] {
    return mappings.map((mapping) => {
      try {
        const rawValue = this.calculateStat(mapping.statId, data);
        let formattedValue = String(rawValue);

        // Apply format if specified
        if (mapping.format) {
          formattedValue = mapping.format.replace("{value}", String(rawValue));
        }

        // Check for change
        const previousValue = previousValues?.get(mapping.obsSourceName);
        const changed = formattedValue !== previousValue;

        return {
          obsSourceName: mapping.obsSourceName,
          value: formattedValue,
          previousValue,
          changed,
        };
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        console.warn(
          `[STATS] Failed to calculate mapping for ${mapping.obsSourceName} (stat: ${mapping.statId}): ${errorMessage}`
        );
        // Return a placeholder value to avoid breaking the array
        return {
          obsSourceName: mapping.obsSourceName,
          value: "ERROR",
          previousValue: previousValues?.get(mapping.obsSourceName),
          changed: true,
        };
      }
    });
  }

  /**
   * Calculate all configured computed stats (arithmetic expressions).
   * Returns the same CalculatedStat shape as calculateMappings, so results
   * can be merged and pushed to OBS identically.
   */
  calculateComputed(
    computed: ComputedStat[],
    data: DndBeyondCharacterResponse,
    previousValues?: Map<string, string>
  ): CalculatedStat[] {
    const variables = getExpressionVariables(data);

    return computed.map((stat) => {
      try {
        const rawValue = evaluateExpression(stat.expression, variables);
        let formattedValue = String(rawValue);

        if (stat.format) {
          formattedValue = stat.format.replace("{value}", String(rawValue));
        }

        const previousValue = previousValues?.get(stat.obsSourceName);
        const changed = formattedValue !== previousValue;

        return {
          obsSourceName: stat.obsSourceName,
          value: formattedValue,
          previousValue,
          changed,
        };
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        console.warn(
          `[STATS] Failed to evaluate computed stat "${stat.id}" (${stat.expression}) for ${stat.obsSourceName}: ${errorMessage}`
        );
        return {
          obsSourceName: stat.obsSourceName,
          value: "ERROR",
          previousValue: previousValues?.get(stat.obsSourceName),
          changed: true,
        };
      }
    });
  }

  /**
   * Get all available stat IDs
   */
  getAvailableStats(): StatId[] {
    return Object.keys(statDefinitions) as StatId[];
  }

  /**
   * Get stat definition by ID
   */
  getStatDefinition(statId: StatId) {
    return statDefinitions[statId];
  }
}

File: rust-src/stats/calculator.rs
use std::collections::HashMap;

use crate::types::DndBeyondCharacterResponse;

use super::definitions::get_stat_definitions;
use super::types::{CalculatedStat, StatMapping};

/// Calculator for D&D Beyond stats
pub struct StatCalculator;

impl StatCalculator {
    pub fn new() -> Self {
        Self
    }

    /// Calculate a single stat value
    pub fn calculate_stat(
        &self,
        stat_id: &str,
        data: &DndBeyondCharacterResponse,
    ) -> Option<String> {
        let definitions = get_stat_definitions();
        let definition = definitions.iter().find(|d| d.id == stat_id)?;
        Some((definition.calculate)(data).to_string())
    }

    /// Calculate all configured stat mappings
    pub fn calculate_mappings(
        &self,
        mappings: &[StatMapping],
        data: &DndBeyondCharacterResponse,
        previous_values: &HashMap<String, String>,
    ) -> Vec<CalculatedStat> {
        mappings
            .iter()
            .map(|mapping| {
                match self.calculate_stat(&mapping.stat_id, data) {
                    Some(raw_value) => {
                        let formatted = if let Some(ref format) = mapping.format {
                            format.replace("{value}", &raw_value)
                        } else {
                            raw_value
                        };

                        let previous = previous_values.get(&mapping.obs_source_name).cloned();
                        let changed = previous.as_ref() != Some(&formatted);

                        CalculatedStat {
                            obs_source_name: mapping.obs_source_name.clone(),
                            value: formatted,
                            previous_value: previous,
                            changed,
                        }
                    }
                    None => {
                        eprintln!(
                            "[STATS] Failed to calculate mapping for {} (stat: {})",
                            mapping.obs_source_name, mapping.stat_id
                        );
                        CalculatedStat {
                            obs_source_name: mapping.obs_source_name.clone(),
                            value: "ERROR".to_string(),
                            previous_value: previous_values
                                .get(&mapping.obs_source_name)
                                .cloned(),
                            changed: true,
                        }
                    }
                }
            })
            .collect()
    }
}
